### PR TITLE
Extra HTTP headers can be passed with CreateOptions

### PR DESCRIPTION
**Pull request by rgba**
- fixes Sending Authentication/token Info #142
- fixes Add extra http header #198

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -122,6 +122,19 @@ async function beforeNavigate(options: CreateOptions, client: any): Promise<void
       options._navigateFailed = true;
     }
   });
+
+  if (options.extraHTTPHeaders){
```
Comment (westy92) [inline src/index.ts]: nit: space before curly brace

File: src/index.ts [hunk 2]
```diff
@@ -122,6 +122,19 @@ async function beforeNavigate(options: CreateOptions, client: any): Promise<void
...
+    // function which reduces the extra http headers array to an object
+    let convertHeaders = (array, keyField, valueField) => {
+      return array.reduce((obj, item) => {
+        obj[item[keyField]] = item[valueField];
+        return obj
+      }, {});
+    }
+
+    Network.setExtraHTTPHeaders({headers : convertHeaders(options.extraHTTPHeaders, 'name','value')})
```
Comment (westy92) [inline src/index.ts]: nit: space after comma

File: test/index.ts [hunk 3]
```diff
@@ -121,6 +121,23 @@ describe('HtmlPdf', () => {
...
+          {
+            name: 'X-Custom-Test-Header',
+            value: 'Passed!'
+          },
+        ],
+      };
+
+      const result = await HtmlPdf.create('http://httpbin.org/headers', options);
+      const pdf = await getParsedPdf(result.toBuffer());
+      expect(pdf.getRawTextContent()).to.contains('X-Custom-Test-Header');
```
Comment (westy92) [inline test/index.ts]: nit: use `contain` instead of `contains` for these two lines

File: src/CreateOptions.ts [hunk 4]
```diff
@@ -93,6 +94,14 @@ export interface CreateOptions {
    */
   cookies?: SetCookieOptions[];
 
+  /**
+   * Extra HTTP headers to send when making a request.
+   *
+   * @type {ExtraHTTPHeader[]}
+   * @memberof CreateOptions
+   */
+  extraHTTPHeaders?: ExtraHTTPHeader[];
```
Comment (westy92) [inline src/CreateOptions.ts]: I'm fine with this approach, but I'd almost prefer specifying them directly as a `string:string` mapping. Then we can avoid the conversion later and prevent introducing the `ExtraHTTPHeader` type into the Chrome namespace, which I'm trying to just have contain classes they define here https://chromedevtools.github.io/devtools-protocol/tot/Network#type-Headers.
i.e. 
```typescript
extraHTTPHeaders: { [key: string]: string; };
```

File: test/index.ts [hunk 5]
```diff
@@ -121,6 +121,23 @@ describe('HtmlPdf', () => {
...
+        port,
+        extraHTTPHeaders: [
+          {
+            name: 'X-Custom-Test-Header',
+            value: 'Passed!'
+          },
+        ],
+      };
+
+      const result = await HtmlPdf.create('http://httpbin.org/headers', options);
```
Comment (westy92) [inline test/index.ts]: 💯 Thanks so much for writing this test!

**Comment by westy92** [comment]
Thank you so much for your contribution! I left a few comments.

File: test/index.ts (hunk 3 above)
Comment (westy92) [reply test/index.ts]: nit: I think `pdf.getRawTextContent()` is somewhat expensive. Call it once, store the response in a variable, and then use it in these two `expect`s.

File: src/index.ts (hunk 1 above)
Comment (rgba) [reply src/index.ts]: fixed

File: src/index.ts (hunk 2 above)
Comment (rgba) [reply src/index.ts]: Refactored code to no longer need the additional data mapping

File: test/index.ts (hunk 3 above)
Comment (rgba) [reply test/index.ts]: fixed both issues

File: src/CreateOptions.ts (hunk 4 above)
Comment (rgba) [reply src/CreateOptions.ts]: Thank you very much for providing more details, this is very much appreciated. I've changed to code to your suggestion, as this is a much simpler and cleaner way of doing this.

File: test/index.ts (hunk 5 above)
Comment (rgba) [reply test/index.ts]: Glad you like it, in my last commit I've enhanced the test case to check for multiple headers.

**Comment by rgba** [comment]
@westy92 Thank you very much for the detailed code review, I think the code should be in good shape now :)

**Comment by westy92** [comment]
I just published this with [v0.6.0](https://github.com/westy92/html-pdf-chrome/releases/tag/v0.6.0).

**Comment by rgba** [comment]
@westy92 Awesome! Thank you very much!

**Pull request merged**